Series of small WCOSS2 updates - round 2

KateFriedman-NOAA opened the pull request:
This PR contains the following:

1. WCOSS2.env updates based on testing and feedback from developers thus far
2. update to wafsgcip rocoto job script to source the env file
3. checkout updates for the GLDAS and WAFS versions, as well as converting UPP GTG checkout to use the new submodules (used to be manage_externals)
4. build updates for gaussian_sfcanl and compile_gfs_utils (includes change to machine-setup.sh for sourcing version files)
5. updates to fcst/efcs configs to set OUTPUT_FILETYPES on WCOSS2 and include EXP_WARM_START variable when adjusting restart_interval for cold-starting
6. remove duplicate DATA removal from wavepost rocoto job scripts

The next PRs will mostly cover workflow updates and resource optimization (depending on testing status).

Refs: #399

KateFriedman-NOAA committed:
Update to checkout script

- update GLDAS tag to gldas_gfsv16_release.v1.20.0
- update WAFS tag to gfs_wafs.v6.2.1
- update UPP checkout to remove manage_externals for GTG part
- add new gtg_git_args variable when -o flag is passed; will recursively
clone the UPP to also obtain the GTG submodule

Refs: #399

KateFriedman-NOAA committed:
Update to gaussian_sfcanl makefile

- change "-L${HDF5}/lib" to "-L${HDF5_LIBRARIES}"
- HDF5 variable no longer set in module, use HDF5_LIBRARIES instead

Refs: #399

KateFriedman-NOAA committed:
Updates for compile_gfs_util build script

- add HOMEgfs variables to both machine-setup.sh and
compile_gfs_util_wcoss.sh to support ingesting the new version files at
build time

Refs: #399

KateFriedman-NOAA committed:
Add env file sourcing in wafsgcip job script

- add sourcing of env file to get MPIRUN variable needed by job at
runtime

Refs: #399

KateFriedman-NOAA committed:
WCOSS2 env file updates from testing

- adjust base MPI settings, add FI_OFI_RXM_*X_SIZE variables for the
forecast jobs, and add specific settings for anal jobs
- update APRUN_GSI, APRUN_ENKF, APRUN_FV3 launcher settings
- add wafsgcip block and MPIRUN setting to support wafsgcip job

Refs: #399

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:feature/ops-wcoss2' into feature/ops-wcoss2

KateFriedman-NOAA committed:
Update forecast configs for WCOSS2

- add WCOSS2 blocks to set OUTPUT_FILETYPES
- add EXP_WARM_START to help properly set restart_interval

Refs: #399

aerorahul commented on env/WCOSS2.env:
```diff
@@ -43,11 +43,29 @@ elif [ $step = "waveinit" -o $step = "waveprep" -o $step = "wavepostsbs" -o $ste
 elif [ $step = "anal" ]; then
 
+    # OMP settings
+    export OMP_PLACES=cores
+    export OMP_NUM_THREADS=$nth_anal
+    export MP_EAGER_LIMIT=165536
+    export MP_COREFILE_FORMAT=lite
+    export MP_EUIDEVELOP=min
+    export MP_EUIDEVICE=sn_all
+    export MP_EUILIB=us
+    export MP_MPILIB=mpich2
+    export MP_LABELIO=yes
+    export MP_SINGLE_THREAD=yes
+    export MP_USE_BULK_XFER=yes
+    export MP_SHARED_MEMORY=yes
+    #export MPICH_ALLTOALL_THROTTLE=0
+    export MP_COLLECTIVE_OFFLOAD=no
+    export KMP_STACKSIZE=1024m
+    export FI_OFI_RXM_SAR_LIMIT=3145728
```
If this is used in `anal.sh`, it might be useful in `eobs.sh` too?  @RussTreadon-NOAA thoughts?

RussTreadon-NOAA replied:
I don't know why all these settings are here for WCOSS2.  Most of these are not in my j-job drivers.    Are these settings from @GeorgeVandenberghe-NOAA  or @MichaelLueken-NOAA

KateFriedman-NOAA replied:
@RussTreadon-NOAA I took those settings from @MichaelLueken-NOAA's test scripts. I can adjust/remove these if they aren't needed...will need to retest analysis job without them to be sure.

MichaelLueken replied:
@KateFriedman-NOAA and @RussTreadon-NOAA, I made changes to my test script to assist the RTMA/URMA port.  You should use:

/lfs/h2/emc/da/noscrub/Michael.Lueken/GSI/scripts/rungsi1534L127_4denvar.sh_orig

or

export OMP_PLACES=cores
export OMP_NUM_THREADS=$threads
export OMP_STACKSIZE=1G
export FI_OFI_RXM_SAR_LIMIT=3145728

the rest aren't necessary.

KateFriedman-NOAA replied:
Thanks @MichaelLueken-NOAA, I have updated the anal block in WCOSS2.env based on your recommendations. Do any other DA jobs need those or similar settings for runtime? Thanks!

MichaelLueken replied:
@KateFriedman-NOAA, As far as I'm aware, those four environment variables should be all you need to run the GSI components.  The necessary environment variables for the EnKF are properly set for those components.

KateFriedman-NOAA replied:
Roger that, thanks @MichaelLueken-NOAA!

aerorahul commented on parm/config/config.efcs:
```diff
@@ -56,15 +56,18 @@ export restart_interval=${restart_interval:-6}
     export  ichunk3d=0; export jchunk3d=0;  export kchunk3d=0
     RESTILE=`echo $CASE_ENKF |cut -c 2-`
-    if [[ "$machine" == "WCOSS_DELL_P3" ]] || [[ "$machine" == "WCOSS2" ]]; then
+    if [[ "$machine" == "WCOSS2" ]]; then
```
This section could use a bit of clean-up.  There is just too much duplication in the if-blocks that can be pulled out.

KateFriedman-NOAA replied:
cleaned up and consolidated, please see updates

aerorahul commented on parm/config/config.fcst:
```diff
@@ -155,7 +155,14 @@ if [[ "$OUTPUT_FILE" == "netcdf" ]]; then
     export  ichunk3d=0; export jchunk3d=0;  export kchunk3d=0
     RESTILE=`echo $CASE |cut -c 2-`
-    if [[ "$machine" == "WCOSS_DELL_P3" ]] || [[ "$machine" == "WCOSS2" ]]; then
+    if [[ "$machine" == "WCOSS2" ]]; then
```
same here.

KateFriedman-NOAA replied:
cleaned up and consolidated, please see updates

aerorahul commented on sorc/machine-setup.sh:
```diff
@@ -16,6 +16,8 @@ else
 fi
 
+HOMEgfs=${HOMEgfs:-`pwd`/../}
+echo "HOMEgfs=$HOMEgfs"
```
please remove.

KateFriedman-NOAA replied:
The echo (definitely can go, will remove) or line 19 too? Line 19 supports the change on line 214: 
`. ${HOMEgfs}/versions/build.ver`
This helps the compile_gfs_util_wcoss.sh script interface with machine-setup.sh since it's in a different folder. Happy to go over the issue and consider alternative fixes.

KateFriedman-NOAA replied:
echo line removed

KateFriedman-NOAA replied:
Any further comments/change requests on this bit?

aerorahul replied:
Line 19 is fine for now.

aerorahul reviewed (commented):
some comments for clean-up.

KateFriedman-NOAA committed:
Remove temporary echo in machine-setup.sh

Refs: #399

GeorgeVandenberghe-NOAA commented:
Some of these date to the old CCS.  I don't think they came from me.   I
tend to fork my own simpler scripts and report results but not commit.


On Wed, Dec 8, 2021 at 3:51 PM RussTreadon-NOAA ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In env/WCOSS2.env
> <https://github.com/NOAA-EMC/global-workflow/pull/523#discussion_r765236325>
> :
>
> > +    export OMP_PLACES=cores
> +    export OMP_NUM_THREADS=$nth_anal
> +    export MP_EAGER_LIMIT=165536
> +    export MP_COREFILE_FORMAT=lite
> +    export MP_EUIDEVELOP=min
> +    export MP_EUIDEVICE=sn_all
> +    export MP_EUILIB=us
> +    export MP_MPILIB=mpich2
> +    export MP_LABELIO=yes
> +    export MP_SINGLE_THREAD=yes
> +    export MP_USE_BULK_XFER=yes
> +    export MP_SHARED_MEMORY=yes
> +    #export MPICH_ALLTOALL_THROTTLE=0
> +    export MP_COLLECTIVE_OFFLOAD=no
> +    export KMP_STACKSIZE=1024m
> +    export FI_OFI_RXM_SAR_LIMIT=3145728
>
> I don't know why all these settings are here for WCOSS2. Most of these are
> not in my j-job drivers. Are these settings from @GeorgeVandenberghe-NOAA
> <https://github.com/GeorgeVandenberghe-NOAA> or @MichaelLueken-NOAA
> <https://github.com/MichaelLueken-NOAA>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/global-workflow/pull/523#discussion_r765236325>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANDS4FSXSD6CLTFGAYHYYLTUP7AN7ANCNFSM5JUVQVRA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 

George W Vandenberghe

*IMSG* at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2141

College Park, MD 20740

***@***.***

301-683-3769(work)  3017751547(cell)

KateFriedman-NOAA committed:
Adjust anal MPI settings in WCOSS2 env file

- remove some unneeded settings and add OMP_STACKSIZE based on review by
DA folks

Refs: #399

RussTreadon-NOAA commented:
Please note the following entry from the WCOSS2 slack channel made by Jim Abeles
```
Jim Abeles 12:51 PM (Oct 22) 
Arun asked me to post how to use mpicexec when using threads:
export OMP_PROC_BIND=true
export OMP_STACKSIZE={the value you need}
export FOR_DISABLE_KMP_MALLOC=TRUE
mpiexec -n {number of tasks} -ppn {tasks per node} --cpu-bind core --depth {number of OMP_NUM_THREADS} {exec}
The above is if your job is not useing hyperthreads
```
These are what I used in the DA j-jobs drivers on WCOSS2.   In addition I set two other environment variables
```
# Set variable to prevent mpi collective hangs
export FI_OFI_RXM_SAR_LIMIT=3145728

# Set variable to get reproducible global_enkf.x
export MPICH_COLL_OPT_OFF=1
```
I recommend cross checking with Jim Abeles and George Vandenberghe to see if the above are still the recommended environment variables to set for threaded mpi applications.

KateFriedman-NOAA commented:
@RussTreadon-NOAA Thanks, I'll tag up with Jim/George to double check on the thread settings. I think I have it set correct at the moment for the DA jobs, I ran with these settings when invoking global_gsi.x (APRUN_GSI) or global_enkf.x (APRUN_ENKF):
```
env file setting:
APRUN_GSI="$launcher -n ${npe_gsi} -ppn $npe_node_anal --cpu-bind core --depth $nth_anal
APRUN_GSI="$launcher -n ${npe_gsi} -ppn $npe_node_eobs --cpu-bind core --depth $nth_eobs

became:
anal: mpiexec -l -n 768 -ppn 16 --cpu-bind core --depth 8 global_gsi.x
eobs: mpiexec -l -n 480 -ppn 64 --cpu-bind core --depth 2 global_gsi.x
```
```
env file setting:
APRUN_ENKF="$launcher -n ${npe_eupd} -ppn $npe_node_eupd --cpu-bind core --depth $nth_eupd

became:
eupd: mpiexec -l -n 848 -ppn 8 --cpu-bind core --depth 16 global_enkf.x
```

Thanks also for your other two settings. I have this in WCOSS2.env for the eupd job (global_enkf.x):
```
    export OMP_PLACES=cores
    export OMP_NUM_THREADS=$nth_eupd
    export OMP_STACKSIZE=2G
    export MPICH_COLL_OPT_OFF=1
    export FI_OFI_RXM_SAR_LIMIT=3145728
```

...and now have this for the anal job:
```
    export OMP_PLACES=cores
    export OMP_NUM_THREADS=$nth_anal
    export OMP_STACKSIZE=1G
    export FI_OFI_RXM_SAR_LIMIT=3145728
```
...and eobs job:
```
    export OMP_PLACES=cores
    export OMP_NUM_THREADS=$nth_eobs
    export OMP_STACKSIZE=1G
    export FI_OFI_RXM_SAR_LIMIT=3145728
```

KateFriedman-NOAA committed:
Add MPI settings to WCOSS2.env for eobs

- added MPI settings to eobs block of WCOSS2.env
- same settings as from anal block (for global_gsi.x)

Refs: #399

GeorgeVandenberghe-NOAA commented:
These look good to me.   A reminder

export FI_OFI_RXM_SAR_LIMIT=3145728

*prevents hangs and THAT'S  A SOLID BUG IN THE CRAY MPI
IMPLEMENTATION.   The setting should affect performance, but not cause
program failure.*




On Thu, Dec 9, 2021 at 10:16 AM Kate Friedman ***@***.***>
wrote:

> @RussTreadon-NOAA <https://github.com/RussTreadon-NOAA> Thanks, I'll tag
> up with Jim/George to double check on the thread settings. I think I have
> it set correct at the moment for the DA jobs, I ran with these settings
> when invoking global_gsi.x (APRUN_GSI) or global_enkf.x (APRUN_ENKF):
>
> env file setting:
> APRUN_GSI="$launcher -n ${npe_gsi} -ppn $npe_node_anal --cpu-bind core --depth $nth_anal
> APRUN_GSI="$launcher -n ${npe_gsi} -ppn $npe_node_eobs --cpu-bind core --depth $nth_eobs
>
> became:
> anal: mpiexec -l -n 768 -ppn 16 --cpu-bind core --depth 8 global_gsi.x
> eobs: mpiexec -l -n 480 -ppn 64 --cpu-bind core --depth 2 global_gsi.x
>
> env file setting:
> APRUN_ENKF="$launcher -n ${npe_eupd} -ppn $npe_node_eupd --cpu-bind core --depth $nth_eupd
>
> became:
> eupd: mpiexec -l -n 848 -ppn 8 --cpu-bind core --depth 16 global_enkf.x
>
> Thanks also for your other two settings. I have this in WCOSS2.env for the
> eupd job (global_enkf.x):
>
>     export OMP_PLACES=cores
>     export OMP_NUM_THREADS=$nth_eupd
>     export OMP_STACKSIZE=2G
>     export MPICH_COLL_OPT_OFF=1
>     export FI_OFI_RXM_SAR_LIMIT=3145728
>
> ...and now have this for the anal job:
>
>     export OMP_PLACES=cores
>     export OMP_NUM_THREADS=$nth_anal
>     export OMP_STACKSIZE=1G
>     export FI_OFI_RXM_SAR_LIMIT=3145728
>
> ...and eobs job:
>
>     export OMP_PLACES=cores
>     export OMP_NUM_THREADS=$nth_eobs
>     export OMP_STACKSIZE=1G
>     export FI_OFI_RXM_SAR_LIMIT=3145728
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/global-workflow/pull/523#issuecomment-989946925>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANDS4FUNYCWFQGGDU2XRQKLUQDB5LANCNFSM5JUVQVRA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 

George W Vandenberghe

*IMSG* at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2141

College Park, MD 20740

***@***.***

301-683-3769(work)  3017751547(cell)

KateFriedman-NOAA committed:
Add missing thread setting to APRUN env settings

- adjust APRUN settings in WCOSS2.env to add needed flags that were
missing
- add "--cpu-bind core --depth $NTHREADS" where needed

Refs: #399

KateFriedman-NOAA committed:
Resource config updates

- add ntasks=$npe_analcalc for analcalc scripts to get
- add npe_node_cycle to jobs that need it; supports changes in env file
- add memory_postsnd and memory_analcalc

Refs: #399

KateFriedman-NOAA commented:
I updated WCOSS2.env to add missing settings for jobs that use threads. Committed @ https://github.com/NOAA-EMC/global-workflow/pull/523/commits/b554ebeef7b4ce66d53301df8e74c367027e815a. Please take a look @GeorgeVandenberghe-NOAA , thanks!

Also updated the config.resources files to add npe_node_cycle and to commit a couple fixes I'd left out of this PR. Committed @ https://github.com/NOAA-EMC/global-workflow/pull/523/commits/433323d0884fc020da1e45e5961a3e98953703f0.

KateFriedman-NOAA committed:
Update UPP checkout to upp_v8.1.0 tag

Refs: #399

KateFriedman-NOAA committed:
Remove unneeded DATA cleanup in wavepost scripts

- the JJOB scripts handled DATA cleanup so remove unneeded duplicate
cleanup from rocoto job scripts
- resolves error at end of job

Refs: #399

GeorgeVandenberghe-NOAA commented:
Send me the stdouts and logs from your slow runs of the GFS.  For a quick
look at how a run is doing check for 'phase 1' in the stdout.  Most of
these should
be around 0.5 seconds.   Consistently over 0.6 indicates a problem. Also
netcdf write times should be under 100 seconds per file.

The pathology I have seen with compression disabled and netccdf-parallel
enabled, is that these netcdf write times go over 300 seconds per file and
the I/O processors back up.





On Mon, Dec 13, 2021 at 10:36 AM Kate Friedman ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In sorc/machine-setup.sh
> <https://github.com/NOAA-EMC/global-workflow/pull/523#discussion_r767870176>
> :
>
> > @@ -16,6 +16,8 @@ else
>      __ms_shell=sh
>  fi
>
> +HOMEgfs=${HOMEgfs:-`pwd`/../}
> +echo "HOMEgfs=$HOMEgfs"
>
> Any further comments/change requests on this bit?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/global-workflow/pull/523#discussion_r767870176>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANDS4FXTIQVQQZJQIWLG573UQYHGTANCNFSM5JUVQVRA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 

George W Vandenberghe

*IMSG* at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2141

College Park, MD 20740

***@***.***

301-683-3769(work)  3017751547(cell)

KateFriedman-NOAA committed:
Consolidate OUTPUT_FILETYPES settings in configs

- cleanup and consolidate the OUTPUT_FILETYPES blocks in config.fcst and
config.efcs

Refs: #399

KateFriedman-NOAA commented:
@aerorahul I cleaned up the OUTPUT_FILETYPES blocks in config.fcst and config.efcs. Please let me know if I have addressed your comments and if there are any others before I submit this PR. Thanks!

GeorgeVandenberghe-NOAA commented:
THe log in  ./comrot/testcyc768e/logs/2021111100/gfsfcst.log
/lfs/h2/emc/ptmp/Kate.Friedman

shows very slow integration times, about 4x slower than would be expected
with 8 threads and ppn=16.   Make sure you are getting exclusive node use.

Otherwise I can't see why yours is so slow and mine isn't.    Also check at
some point to get the number of ranks per I/O  groiyp  down to 64 but I
don't think
this what's slowing it down.



On Mon, Dec 13, 2021 at 11:32 AM Kate Friedman ***@***.***>
wrote:

> @aerorahul <https://github.com/aerorahul> I cleaned up the
> OUTPUT_FILETYPES blocks in config.fcst and config.efcs. Please let me know
> if I have addressed your comments and if there are any others before I
> submit this PR. Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/global-workflow/pull/523#issuecomment-992655780>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANDS4FVUG6XIKVN3Z37BVXLUQYN23ANCNFSM5JUVQVRA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 

George W Vandenberghe

*IMSG* at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2141

College Park, MD 20740

***@***.***

301-683-3769(work)  3017751547(cell)

aerorahul commented on env/WCOSS2.env:
```diff
@@ -86,40 +91,53 @@ elif [ $step = "gldas" ]; then
     export NTHREADS_ENKF=${nth_eupd:-$nth_max}
     [[ $NTHREADS_ENKF -gt $nth_max ]] && export NTHREADS_ENKF=$nth_max
-    export APRUN_ENKF="$launcher -n ${npe_enkf:-${npe_eupd:-$PBS_NP}} -ppn $npe_node_eupd"
+    export APRUN_ENKF="$launcher -n ${npe_enkf:-${npe_eupd:-$PBS_NP}} -ppn $npe_node_eupd --cpu-bind core --depth $nth_eupd"
```
```suggestion
    export APRUN_ENKF="$launcher -n ${npe_enkf:-${npe_eupd:-$PBS_NP}} -ppn $npe_node_eupd --cpu-bind core --depth $NTHREADS_ENKF"
```

aerorahul commented on env/WCOSS2.env:
```diff
@@ -86,40 +91,53 @@ elif [ $step = "gldas" ]; then
     export NTHREADS_GSI=${nth_eobs:-$nth_max}
     [[ $NTHREADS_GSI -gt $nth_max ]] && export NTHREADS_GSI=$nth_max
-    export APRUN_GSI="$launcher -n ${npe_gsi:-${npe_eobs:-$PBS_NP}} -ppn $npe_node_eobs"
+    export APRUN_GSI="$launcher -n ${npe_gsi:-${npe_eobs:-$PBS_NP}} -ppn $npe_node_eobs --cpu-bind core --depth $nth_eobs"
```
```suggestion
    export APRUN_GSI="$launcher -n ${npe_gsi:-${npe_eobs:-$PBS_NP}} -ppn $npe_node_eobs --cpu-bind core --depth $NTHREADS_GSI"
```

aerorahul commented on env/WCOSS2.env:
```diff
@@ -76,7 +81,7 @@ elif [ $step = "gldas" ]; then
     export NTHREADS_GLDAS=${nth_gldas:-$nth_max}
     [[ $NTHREADS_GLDAS -gt $nth_max ]] && export NTHREADS_GLDAS=$nth_max
-    export APRUN_GLDAS="$launcher -n $npe_gldas -ppn $npe_node_gldas"
+    export APRUN_GLDAS="$launcher -n $npe_gldas -ppn $npe_node_gldas --cpu-bind core --depth $nth_gldas"
```
```suggestion
    export APRUN_GLDAS="$launcher -n $npe_gldas -ppn $npe_node_gldas --cpu-bind core --depth $NTHREADS_GLDAS"
```

aerorahul commented on env/WCOSS2.env:
```diff
@@ -86,40 +91,53 @@ elif [ $step = "gldas" ]; then
 elif [ $step = "eobs" ]; then
 
+    export OMP_PLACES=cores
+    export OMP_NUM_THREADS=$nth_eobs
```
Why is this being set here?  This should be set in the ex-script immediately prior to the execution.
Also, what if `nth_eobs` is not the correct one?  `NTHREADS_GSI` is the correct one that is being supplied.

KateFriedman-NOAA replied:
I have corrected the NTHREADS settings (other comment) and removed the unneeded OMP_NUM_THREADS line.

Was there a question about OMP_PLACES too? That came from discussions with @RussTreadon-NOAA , @MichaelLueken-NOAA , and @GeorgeVandenberghe-NOAA .

aerorahul commented on env/WCOSS2.env:
```diff
@@ -86,40 +91,53 @@ elif [ $step = "gldas" ]; then
 elif [ $step = "eupd" ]; then
 
+    export OMP_PLACES=cores
+    export OMP_NUM_THREADS=$nth_eupd
```
Why is this being set here?  This should be set in the ex-script immediately prior to the execution.
Also, what if `nth_eupd` is not the correct one?  `NTHREADS_ENKF` is the correct one that is being supplied.

KateFriedman-NOAA replied:
I have corrected the NTHREADS settings (other comment) and removed the unneeded OMP_NUM_THREADS line.

Is there a question about OMP_PLACES?

aerorahul replied:
No question about `OMP_PLACES`.

aerorahul commented on parm/config/config.fcst:
```diff
@@ -155,26 +155,17 @@ if [[ "$OUTPUT_FILE" == "netcdf" ]]; then
     RESTILE=`echo $CASE |cut -c 2-`
-    if [[ "$machine" == "WCOSS_DELL_P3" ]] || [[ "$machine" == "WCOSS2" ]]; then
+    if [ $RESTILE -ge 768 ]; then
+        export OUTPUT_FILETYPES=" 'netcdf_parallel' 'netcdf_parallel' "
```
I don't think this is just for C768 or greater.  Looking at the left, this is being set for HERA and ORION for all resolutions. except less than (or equal) to C192.

KateFriedman-NOAA replied:
Have adjusted the OUTPUT_FILETYPES settings again, please review new changes in config.fcst.

aerorahul reviewed (changes requested):
Please see comments.

KateFriedman-NOAA committed:
Adjust OUTPUT_FILETYPES in config.fcst

- readjust the OUTPUT_FILETYPES settings in config.fcst
- added Hera and Orion blocks back in

Refs: #399

KateFriedman-NOAA committed:
Change nth_gldas to NTHREADS_GLDAS

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

KateFriedman-NOAA committed:
Change nth_eobs to NTHREADS_GSI

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

KateFriedman-NOAA committed:
Change nth_eupd to NTHREADS_ENKF

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

KateFriedman-NOAA committed:
Add missing mpiexec flag for fcst jobs

- add missing "--cpu-bind core" with APRUN_FV3 launcher settings
- model runs 4x slower without

Refs: #399

KateFriedman-NOAA committed:
Remove unneeded OMP_NUM_THREADS in WCOSS2.env

- remove unneeded OMP_NUM_THREADS in WCOSS2 env file for the anal, eobs,
and eupd jobs

Refs: #399

aerorahul commented on env/WCOSS2.env:
```diff
@@ -86,40 +90,51 @@ elif [ $step = "gldas" ]; then
+      export APRUN_FV3="$launcher -n ${npe_fcst:-$PBS_NP} -ppn $npe_node_fcst --cpu-bind core --depth $NTHREADS_FV3"
     else
-      export APRUN_FV3="$launcher -n ${npe_fcst_gfs:-$PBS_NP} -ppn $npe_node_fcst"
+      export APRUN_FV3="$launcher -n ${npe_fcst_gfs:-$PBS_NP} -ppn $npe_node_fcst --cpu-bind core --depth $NTHREADS_FV3"
```
Should this be `npe_node_fcst_gfs`?

KateFriedman-NOAA replied:
One would think so but npe_node_fcst_gfs isn't a variable that exists in the system. It gets calculated for the gfsfcst job through the configs. For the gfsfcst job the following happens:
We start with config.fcst which immediately sources these for the gfsfcst job from config.fv3:
```
        export layout_x_gfs=16
        export layout_y_gfs=16
        export npe_wav_gfs=440
        export nth_fv3_gfs=7
        export WRITE_GROUP_GFS=7
        export WRTTASK_PER_GROUP_GFS=64
```
...then config.fcst sets:
```
[[ "$CDUMP" == "gfs" ]] && export nth_fv3=$nth_fv3_gfs
```
...then config.fcst sources config.resources and sets:
```
    export nth_fcst=${nth_fv3:-2}
    export npe_fcst_gfs=$(echo "$layout_x_gfs * $layout_y_gfs * 6" | bc)
    export npe_node_fcst=$(echo "$npe_node_max / $nth_fcst" | bc)
```
...then returning to config.fcst we go through the following if-blocks:
```
export npe_fv3=$npe_fcst
if [[ "$CDUMP" == "gfs" ]] ; then
   export npe_fv3=$npe_fcst_gfs
   export layout_x=$layout_x_gfs
   export layout_y=$layout_y_gfs
   export WRITE_GROUP=$WRITE_GROUP_GFS
   export WRTTASK_PER_GROUP=$WRTTASK_PER_GROUP_GFS
fi
if [ $QUILTING = ".true." ]; then
    export npe_fv3=$(echo " $npe_fv3 + $WRITE_GROUP * $WRTTASK_PER_GROUP" | bc)
    export npe_fcst=$npe_fv3
    export npe_fcst_gfs=$(echo " $npe_fcst_gfs  + $WRITE_GROUP_GFS * $WRTTASK_PER_GROUP_GFS" | bc)
fi
if [ $DO_WAVE = "YES" ] ; then
  export npe_fcst=$((npe_fcst + npe_wav))
  if [ "$WAVE_CDUMP" = "gfs" -o "$WAVE_CDUMP" = "both" ]; then
     export npe_fcst_gfs=$((npe_fcst_gfs + npe_wav_gfs))
     if [ "$CDUMP" = "gfs" ]; then npe_wav=$npe_wav_gfs ; fi
  fi
fi
```
If you follow through all of that you get:
```
to start:
        export layout_x_gfs=16
        export layout_y_gfs=16
        export npe_wav_gfs=440
        export nth_fv3_gfs=7
        export WRITE_GROUP_GFS=7
        export WRTTASK_PER_GROUP_GFS=64
..then
        nth_fv3=$nth_fv3_gfs=7
        nth_fcst=$nth_fv3=7
        npe_fcst_gfs=16*16*6=1536
        npe_node_fcst=128/7=~18
...then
        npe_fv3=1536
        layout_x=16
        layout_y=16
        WRITE_GROUP=7
        WRTTASK_PER_GROUP=64
...then
        npe_fcst_gfs=1536 + 7 * 64 = 1984
...then
        npe_fcst_gfs=1984 + 440 = 2424
        npe_wav=440
```

From all of that you get the gfsfcst settings but, to your comment, you get `npe_node_fcst` that is set for the gfs and not the gdas. Thankfully this was all redone for the coupled system (what is now in develop).

aerorahul replied:
Thanks for the explanation.  
We have a tendency to append `_gfs` to the early cycle stuff, so was just checking.

aerorahul reviewed (approved):
Looks good.
I have just 1 clarification.
Check the comment.
